misk-aws/src/main/kotlin/misk/jobqueue/sqs/SqsJobConsumer.kt

```diff
@@ -45,91 +55,121 @@ internal class SqsJobConsumer @Inject internal constructor(
+    private fun receive(): List<Status> {
+      val messages = fetchMessages()
 
       if (messages.size == 0) {
-        return Status.NO_WORK
+        return listOf(Status.NO_WORK)
       }
 
-      val futures = messages.map { SqsJob(queueName, queues, metrics, it) }.map { message ->
-        dispatchThreadPool.submit {
-          metrics.jobsReceived.labels(queueName.value, queueName.value).inc()
+      val futures = handleMessages(messages)
+
+      return CompletableFuture.allOf(*futures.toTypedArray()).thenApply {
+        futures.map { it.join() }
+      }.join()
+    }
+
+
+    private fun handleMessages(messages: List<SqsJob>): List<CompletableFuture<Status>> {
+      return messages.map { message ->
+        CompletableFuture.supplyAsync(Supplier {
+          metrics.jobsReceived.labels(queue.queueName, queue.queueName).inc()
 
-          tracer.traceWithSpan("handle-job-${queueName.value}") { span ->
+          tracer.traceWithSpan("handle-job-${queue.queueName}") { span ->
             // If the incoming job has an original trace id, set that as a tag on the new span.
             // We don't turn that into the parent of the current span because that would
             // incorrectly include the execution time of the job in the execution time of the
-            // action that triggered the job        q
+            // action that triggered the job
             message.attributes[SqsJob.ORIGINAL_TRACE_ID_ATTR]?.let {
               ORIGINAL_TRACE_ID_TAG.set(span, it)
             }
 
             // Run the handler and record timing
             try {
               val (duration, _) = timed { handler.handleJob(message) }
-              metrics.handlerDispatchTime.record(duration.toMillis().toDouble(), queueName.value, queueName.value)
+              metrics.handlerDispatchTime.record(duration.toMillis().toDouble(), queue.queueName, queue.queueName)
+              Status.OK
             } catch (th: Throwable) {
-              log.error(th) { "error handling job from ${queueName.value}" }
-              metrics.handlerFailures.labels(queueName.value, queueName.value).inc()
+              log.error(th) { "error handling job from ${queue.queueName}" }
+              metrics.handlerFailures.labels(queue.queueName, queue.queueName).inc()
               Tags.ERROR.set(span, true)
-              throw th
+              Status.FAILED
             }
           }
-        }
+        }, dispatchThreadPool)
       }
-
-      for (future in futures) {
-        try {
-          future.get()
-        } catch (e: ExecutionException) {
-          // the exception was already logged when the dispatched task failed above
-          return Status.FAILED
-        }
-      }
-
-      return Status.OK
     }
   }
 
   companion object {
     private val log = getLogger<SqsJobConsumer>()
-
+    internal val RECEIVERS_PER_QUEUE = Feature("jobqueue-receiver")
```

Comment:
Originally was called. `jobqueue-consumers`. Should i keep that name or change it to `jobqueue-receivers`?

Reply:
I lean towards "consumer(s)" since this feature controls the number of message consumers (either per node or across the cluster).

If you end up keeping jobqueue-consumers, I'd also go ahead and change the code/var names/etc. to mention "consumers" instead of "receivers" for the sake of legibility/consistency.

Reply:
The original implementation constant is `RECEIVERS_PER_QUEUE_FEATURE` so that's why I used `jobqueue-receiver` feature flag name. So we need to decide if we're consistent between implementations or in with the feature flag name we used 😬

Reply:
Yeah, the original impl uses consumer and receiver interchangeably :\

I'd rather stick to consumer; seems semantically more correct and both the feature flag and leases have "consumer" in their names; "receiver" only really exists in code so it'll be a cosmetic change (I'll open a PR to make it consistent across impls after we get this one merged).